libraries/chain/controller.cpp

```diff
@@ -3999,15 +3986,95 @@ struct controller_impl {
+      EOS_ASSERT( !f_ext.qc_claim.is_strong_qc,
+                  block_validate_exception,
+                  "Transition block #${b} has a strong QC claim",
+                  ("b", block_num) );
+      EOS_ASSERT( !f_ext.new_proposer_policy_diff,
+                  block_validate_exception,
+                  "Transition block #${b} has new_proposer_policy_diff",
+                  ("b", block_num) );
+
+      if (auto it = prev.header_exts.find(finality_extension::extension_id()); it != prev.header_exts.end()) {
+         const auto& prev_finality_ext = std::get<finality_extension>(it->second);
+         EOS_ASSERT( f_ext.qc_claim.block_num == prev_finality_ext.qc_claim.block_num,
+                     block_validate_exception,
+                     "Transition block #${b} QC claim block_num not equal to previous QC claim block_num",
+                     ("b", block_num) );
+      } else {
+         // Savanna Genesis Block
+         EOS_ASSERT( f_ext.qc_claim.block_num == block_num,
+                     block_validate_exception,
+                     "Savanna Genesis block #${b} QC claim block_num not equal to current block_num",
+                     ("b", block_num) );
+         EOS_ASSERT( f_ext.new_finalizer_policy_diff,
+                     block_validate_exception,
+                     "Savanna Genesis block #${b} finality block header extension misses new_finalizer_policy_diff",
+                     ("b", block_num) );
+
+         finalizer_policy no_policy;
+         // apply_diff will FC_ASSERT if new_finalizer_policy_diff is malformated
+         finalizer_policy genesis_policy = no_policy.apply_diff(*f_ext.new_finalizer_policy_diff);
+         EOS_ASSERT( genesis_policy.generation == 1,
+                     block_validate_exception,
+                     "Savanna Genesis block #${b} finalizer policy generation (${g}) not 1",
+                     ("b", block_num)("g", genesis_policy.generation) );
+      }
+   }
+
    // thread safe, expected to be called from thread other than the main thread
    template<typename ForkDB, typename BS>
    block_handle create_block_state_i( ForkDB& forkdb, const block_id_type& id, const signed_block_ptr& b, const BS& prev ) {
       constexpr bool savanna_mode = std::is_same_v<typename std::decay_t<BS>, block_state>;
+
+      // Verify claim made by finality_extension in block header extension and
+      // quorum_certificate_extension in block extension are valid.
+      // This is the only place the evaluation is done.
+      // Note: the purpose of running verify_qc_claim on Legacy blocks too is to
+      // safe guard Legacy blocks.
       if constexpr (savanna_mode) {
-         // Verify claim made by finality_extension in block header extension and
-         // quorum_certificate_extension in block extension are valid.
-         // This is the only place the evaluation is done.
          verify_qc_claim(id, b, prev);
```

Comment:
If `savanna_mode` is true, then it means that the parent block of block `b` has a `block_state`.

For a block to have a `block_state` it means the block must be a Savanna block, which means either Transition block or a Proper Savanna block.

A block that has a Savanna block as a parent must also be a Savanna block. A block that has a Proper Savanna block as a parent must also be a Proper Savanna block.

So given that the parent block has a `block_state`, we know that block `b` must either be a Transition block or a Proper Savanna block.

Perhaps there are some other restrictions in the caller of `create_block_state_i` that guarantee that block `b` will never be a Transition block. But that is not obvious purely from the context in this function.

Regardless, `verify_qc_claim` (as it is currently implemented) is appropriate to call on any Savanna block (including Transition blocks) that has a Savanna block as a parent.

@linh2931: Maybe you could put a comment like the above here to make these clearer for the reader. This will especially be useful when we later refactor how these validations work.

By the way, if we are sure that `create_block_state_i` will never be called with a Transition block `b` and with a `prev` that decays to type `block_state`, then we should assert that prior to the call to `verify_qc_claim`. The assert could simply check that `b` is a Proper Savanna block by ensuring `b->is_proper_svnn_block() == true`.